src/app/InteractionModelEngine.cpp

```diff
@@ -855,6 +855,64 @@ CHIP_ERROR InteractionModelEngine::PushFrontAttributePathList(ObjectList<Attribu
     return err;
 }
 
+void InteractionModelEngine::RemoveDuplicateConcreteAttribute(ObjectList<AttributePathParams> *& aAttributePaths)
```

Comment:
So... this is not implementing the spec proposed in https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/5109, in a number of ways:

1. Per that spec PR, if a request has these two paths:
    * (wildcard endpoint, wildcard cluster, wildcard attribute)
    * (endpoint 1, cluster 1, attribute 1)

   and there is no actual attribute 1 on cluster 1 on endpoint 1, the second path is _not_ discarded and results in an error IB.  That's because "Request Path Expansion" only produces _existent paths_.  But this function will discard that second path.

2. Per spec, if a request has the paths:
    * (wildcard endpoint, wildcard cluster, wildcard attribute)
    * (wildcard endpoint, wildcard cluster, specific attribute)

    then the second path is not discarded, because it is not a concrete path.  But this function will discard it.

3. Per spec, if a request has the paths:
    * (wildcard endpoint, cluster 6, attribute 0)
    * (endpoint 1, cluster 6, attribute 0)

    and endpoint 1 in fact has cluster 6 on it, and that cluster has attribute 0, then the second path will be discarded.  But this function will not discard the second path in this case, because the first path does not have an attribute wildcard.

We need to figure out whether the spec PR is wrong and needs to be adjusted, or whether this code is wrong and needs to be adjusted, or both.  And we should have tests for cases like the above.

Reply:
1. I think current code in reporting engine would not result in an ErrorIB if that attribute does not exists, the non-exist path would be skipped.
// For each path included in the interested path of the read handler...
for (; apReadHandler->GetAttributePathExpandIterator()->Get(readPath);
     apReadHandler->GetAttributePathExpandIterator()->Next())
{
    if (!apReadHandler->IsPriming())
2. I think current code would not discard the second path
while (path1 != nullptr)
{
    // skip all wildcard paths
    if (path1->mValue.HasAttributeWildcard())
    {
        prev  = path1;
        path1 = path1->mpNext;
        continue;
    }
Also I have tested it in unit test.
    attributePathParams1.mEndpointId  = kInvalidEndpointId;
    attributePathParams1.mClusterId   = kInvalidClusterId;
    attributePathParams1.mAttributeId = kInvalidAttributeId;

    attributePathParams2.mEndpointId  = 1;
    attributePathParams2.mClusterId   = kInvalidClusterId;
    attributePathParams2.mAttributeId = kInvalidAttributeId;

    attributePathParams3.mEndpointId  = 1;
    attributePathParams3.mClusterId   = 2;
    attributePathParams3.mAttributeId = kInvalidAttributeId;

    // Wildcards cannot be deduplicated.
    InteractionModelEngine::GetInstance()->PushFrontAttributePathList(attributePathParamsList, attributePathParams1);
    InteractionModelEngine::GetInstance()->PushFrontAttributePathList(attributePathParamsList, attributePathParams2);
    InteractionModelEngine::GetInstance()->PushFrontAttributePathList(attributePathParamsList, attributePathParams3);
    InteractionModelEngine::GetInstance()->RemoveDuplicateConcreteAttribute(attributePathParamsList);
    NL_TEST_ASSERT(apSuite, GetAttributePathListLength(attributePathParamsList) == 3);
    InteractionModelEngine::GetInstance()->ReleaseAttributePathList(attributePathParamsList);

3. I think current code has discarded the second path,
the existing test code has validated this scenario.
attributePathParams1.mEndpointId  = kInvalidEndpointId;
attributePathParams1.mClusterId   = 1;
attributePathParams1.mAttributeId = 1;

attributePathParams2.mEndpointId  = 2;
attributePathParams2.mClusterId   = 1;
attributePathParams2.mAttributeId = 1;

attributePathParams3.mEndpointId  = 3;
attributePathParams3.mClusterId   = 1;
attributePathParams3.mAttributeId = 1;

// 1st path is wildcard endpoint, 2nd, 3rd paths are concrete paths, the concrete ones would be removed.
InteractionModelEngine::GetInstance()->PushFrontAttributePathList(attributePathParamsList, attributePathParams1);
InteractionModelEngine::GetInstance()->PushFrontAttributePathList(attributePathParamsList, attributePathParams2);
InteractionModelEngine::GetInstance()->PushFrontAttributePathList(attributePathParamsList, attributePathParams3);
InteractionModelEngine::GetInstance()->RemoveDuplicateConcreteAttribute(attributePathParamsList);
NL_TEST_ASSERT(apSuite, GetAttributePathListLength(attributePathParamsList) == 1);
InteractionModelEngine::GetInstance()->ReleaseAttributePathList(attributePathParamsList);

Reply:
I think "Request Path Expansion" could be optimized out in spec

Reply:
1.  I don't follow.  For current code, if you try to read (endpoint 1, cluster 1, attribute 1) and there is no such attribute you will absolutely get an error IB.  And per spec that should happen.

2. Ah, it's `HasAttributeWildcard()`, not `HasWildcardAttributeId()`.  That is ... very confusing naming.  And also, it means that the use of `HasAttributeWildcard` in `ClusterStateCache::OnUpdateDataVersionFilterList` is wrong: that callsite wants paths with a wildcard attribute, not with any wildcard.  But yes, this part is ok.

3. Yes, this is the same naming confusion.  `HasAttributeWildcard()` should be named `IsWildcardPath()` to match the spec naming....

Reply:
yep, I would follow-up the IsWildcardPath function name change, and HasWildcardAttributeId function creation for OnUpdateDataVersionFilterList

Reply:
updated for 1, thanks